### ISSUE TITLE
Why won't a getter from a service/state work after any store changes?

### ISSUE THREAD

**Issue by JasonLandbridge** [DanWahlin/Observable-Store#107]
Hi there, 

Just to clarify my understanding: 

If you read: [#using-observable-store-with-angular](https://github.com/DanWahlin/Observable-Store#using-observable-store-with-angular) => Point 9, it says the following: 
```
// Can call service/store to get data directly 
// It won't fire when the store state changes though in this case
//this.storeSub = this.customersService.get().subscribe(custs => this.customers = custs);
```

I find that the following is indeed not working: 

```Typescript
public getIsLoggedIn(): Observable<boolean> {
    return this.stateChanged.pipe(switchMap((x) => of(x.isAuthenticated)));
}
```

**Why isn't it the case that when you call a get wrapper function that it won't fire any subsequent, or even initial/last state changes?**

Also here: [#comment](https://github.com/DanWahlin/Observable-Store/issues/106#issuecomment-774218563) @DanWahlin says the following: 
>  ... Currently the store will send the latest state so I wanted to check on that part of your scenario there.

I'm thinking i'm missing some big part of the puzzle but I haven't been able to figure it out.

Thank you for any answers!

**Comment by spierala** [comment]
How are you using `getIsLoggedIn()`? How do you subscribe? Is the getter maybe called many times?
Another question: Why do you use `switchMap`? `map` should be enough to extract the authenticated value.

**Comment by DanWahlin** [comment]
@JasonLandbridge: Let me know which sample you're referring to. Unless an async pipe is being used (if you're using it with Angular) you should have to subscribe. Either that specific code isn't actually being used or something else is going on there. 

@spierala: I'd also need to know the specific code sample you're looking at, but I actually don't know why switchMap is used there since it doesn't look like there's an inner observable. Might be an artifact from some older code that never got changed.

**Comment by JasonLandbridge** [comment]
Thank you both for the quick answers!

I work in Vue.js with Nuxt.js but I think the following could easily be used in Angular.

So lets say I have two service files:

```Typescript
// authService.ts
export class AuthService extends BaseService {

    constructor() {
        super({
            stateSliceSelector: (state: IStoreState) => {
                return {
                    isAuthenticated: state.isAuthenticated,
                    authToken: state.authToken,
                };
            },
        });
    }

    public getIsLoggedIn(): Observable<boolean> {
        return this.stateChanged.pipe(switchMap((x) => of(x.isAuthenticated)));
    }

    public login(credentials: ICredentials): Observable<boolean> {
        return getAuthToken(credentials).pipe(
            tap((data) => {
                if (data?.user_token !== '') {
                    this.setState({ isAuthenticated: true, authToken: data });
                }
            }),
            switchMap(() => this.getIsLoggedIn()),
            take(1),
        );
    }

    public logout(): Observable<boolean> {
        this.setState({ isAuthenticated: false });
        return this.getIsLoggedIn();
    }
}

const authService = new AuthService();
export default authService;
```

And userService.ts

```Typescript
// userService.ts
export class UserService extends BaseService {
    constructor() {
        super({
            stateSliceSelector: (state: IStoreState) => {
                return {
                    user: state.user,
                };
            },
        });
        AuthService.getIsLoggedIn().subscribe((state) => Log.info('user is logged', state));
        AuthService.getIsLoggedIn()
            // getUserDetails is an observable api call
            .pipe(switchMap(() => getUserDetails()))
            .subscribe((user) => {
                Log.info('user', user[0]);
                this.setState({ user: user[0] });
            });
    }
}

const userService = new UserService();
export default userService;
```


And BaseService is this:

```Typescript
export abstract class BaseService extends ObservableStore<IStoreState> {
	protected constructor(settings: ObservableStoreSettings) {
		settings.trackStateHistory = true;
		super(settings);
	}

	public logHistory(): void {
		Log.warn('history', this.stateHistory);
	}
}
```

> How are you using getIsLoggedIn()? How do you subscribe? Is the getter maybe called many times?

 I'm creating a service layer around my Api calls and observable store, which are then called from my Vue components. I use the constructor of each service class to subscribe to for example getIsLoggedIn to then fire an API call for the user object. Am i wrong to assume that, as long as I'm subscribed to `getIsLoggedIn()` I will continue to receive state changes in other service classes? Or should I use `this.getState.isAuthenticated` everytime I need that value?

> Why do you use switchMap? map should be enough to extract the authenticated value.

Good point! I red several times you should use switchMap for apiCalls (i'm using [axios-observable](https://www.npmjs.com/package/axios-observable)) to switch the return type and I wrongly assumed that would be the way to always do it. Will use map for these kinds of things now!

> Let me know which sample you're referring to. Unless an async pipe is being used (if you're using it with Angular) you should have to subscribe. Either that specific code isn't actually being used or something else is going on there.

I haven't used a sample, mostly because I'm doing it in Vue and that observable store example isn't showing what I was looking for. As shown above, I am indeed subscribing to it but it's never fired even though getIsLoggedIn is changing in the store. My subscription could be set too late, and therefore not receive any statechanges, as a last value might not be given on subscription either. It's funny because I think the solution is staring me right in the face but I just don't see it :P

To summarize, I would like to be able to subscribe once to  `getIsLoggedIn()` from anywhere, and continue to receive stateChanges updates telling me if the user is still logged in and/or do actions when the user is logged out. 

Thanks!

**Comment by spierala** [comment]
So there is not a single subscription on `AuthService.getIsLoggedIn()` which will show an emitted value?
Are the `setState` calls executed as expected?

Your code looks OK at first sight. I created a small example to simulate roughly what you are doing (in Angular though): https://stackblitz.com/edit/angular-test-observable-store?file=src%2Fapp%2Fmy-store.service.ts

You can see in some-other.service.ts that late subscriptions should just work fine (check the console to see the logging).

**Comment by JasonLandbridge** [comment]
@spierala, thanks for the answer!

I managed to solve it but the problem was unrelated to Observable Store. 

I was making separate service classes, and some service classes had a dependency on others. My mistake was putting all those subscriptions in the constructor of each service class, which meant some subscriptions weren't even setup to begin with. 

I solved this by giving each service class a `setup()` function and then call each 'service.setup()' on app startup. 

Now everything works as perfectly!

Stupid mistake on my part.

**Comment by DanWahlin** [comment]
I'm catching up on the thread (thanks for jumping in @spierala  - appreciate that) but am glad you found the root of the problem @JasonLandbridge. I do something similar in my services where I need to initialize something. I call it `init` but it's basically the same purpose as your `setup` call. 

As far as the "stupid mistake on my part" comment, I can sympathize. I've made plenty of those mistakes myself. :-)
